homework/tai_nguyen/lesson8/JS_questions.md

```diff
@@ -0,0 +1,253 @@
+  }
+}
+```
+
+Explanation:
+
+- The `[\-+]?` indicates that number can be positive (with and without `+`) and negative
+- The `(\d(\.?\d*)?|(\.\d+))` indicates that number can be `1.` or `.1` format.
+- The `(([eE][\-+]?)?\d*)?` indicates that number can be presented in format `...e+1` or `E-2`
+
+**Question 3:** Given an array of promises, Promise.all returns a promise that waits for all of the promises in the array to finish. It then succeeds, yielding an array of result values. If a promise in the array fails, the promise returned by all fails too, with the failure reason from the failing promise.
+
+```typescript
+type InferPromiseValue<T> = T extends Promise<infer V> ? V : never;
+type PromiseAll<TPromises extends Promise<any>[]> = InferPromiseValue<
+  TPromises[number]
+>[];
+
+function Promise_all<T extends Promise<any>[]>(promises: T) {
+  return new Promise<PromiseAll<T>>((resolve, reject) => {
+    let resolvedCount = 0;
+    const promiseResults: PromiseAll<T> = [];
+
+    if (promises.length === 0) {
+      resolve(promiseResults);
+    }
+
+    promises.forEach((p, index) => {
+      p.then(
+        function onFullFilled(value) {
+          resolvedCount++;
+          promiseResults[index] = value;
+
+          if (resolvedCount === promises.length) {
+            resolve(promiseResults);
+          }
+        },
+        function onRejected(reason) {
+          reject(reason);
+        }
+      );
+    });
+  });
+}
+```
+
+**Question 4:** How can we log the values that are commented out after the console.log statement?
+
+```javascript
+function* startGame() {
+  const answer = yield 'Do you love JavaScript?';
+  if (answer !== 'Yes') {
+    return "Oh wow... Guess we're gone here";
+  }
+  return 'JavaScript loves you back ❤️';
+}
+
+const game = startGame();
+console.log(/* 1 */); // Do you love JavaScript?
+console.log(/* 2 */); // JavaScript loves you back ❤️
+```
+
+- A: `game.next("Yes").value` `and game.next().value`
+- B: `game.next.value("Yes")` and `game.next.value()`
+- C: `game.next().value` and `game.next("Yes").value`
+- D: `game.next.value()` and `game.next.value("Yes")`
+
+> Answer: TODO
+
+**Question 5:** What's the output?
+
+```javascript
+function getItems(fruitList, ...args, favoriteFruit) {
+  return [...fruitList, ...args, favoriteFruit]
+}
+
+getItems(["banana", "apple"], "pear", "orange")
+```
+
+- A: ["banana", "apple", "pear", "orange"]
+- B: [["banana", "apple"], "pear", "orange"]
+- C: ["banana", "apple", ["pear"], "orange"]
+- D: SyntaxError
+
+> Answer: D
+> The rest operator can only be used at the end of parameters slot of a function
+
+**Question 6:** What's the output?
+
+```javascript
+const getList = ([x, ...y]) => [x, y]
+const getUser = user => { name: user.name, age: user.age } // Syntax Error
+
+const list = [1, 2, 3, 4]
+const user = { name: "Lydia", age: 21 }
+
+console.log(getList(list))
+console.log(getUser(user))
+```
+
+> Answer: Syntax Error
```

Comment:
we have 2 console.log() line

Reply:
Hi @thaycacac, I mean that because of the syntax error, the program will not reach `console.log` 
`Syntax Error` is the output

Reply:
oh, thanks @trongtai37 .  I apologize for miss option this question

```md
•	A: [1, [2, 3, 4]] and SyntaxError
•	B: [1, [2, 3, 4]] and { name: "Lydia", age: 21 }
•	C: [1, 2, 3, 4] and { name: "Lydia", age: 21 }
•	D: Error and { name: "Lydia", age: 21 }
```

Reply:
I think none of them is the answer because the program gets a syntax error.
It will not log.
Do you think so, @thaycacac?

Reply:
yep, i think so =)) but that is options for question. anyway, we should skip this question.